### ISSUE TITLE
Watchman memory usage issue

### ISSUE THREAD

**Issue by arashsa** [facebook/watchman#324]
I'm on Mac OS X 10.11.5. Watchman is using a constant of 2.29 RAM when it is idle. I had an issue with it hovering around 80-100% cpu usage as well, which I fixed per this issue #212.

**Comment by ggamel** [comment]
I've been experiencing this exact issue under 10.11.5, with the exception being watchman will slam the CPU at ~228%+ for 10-20 minutes. Forcing the process to quit does not matter, as it instantly respawns (assuming it quits at all) and cranks the CPU back up.

Not sure what can be done, but having this never happen again would be great. 👍 

watchman version is 4.6.0

**Comment by wez** [comment]
Please provide some more details:

1. The output from `watchman watch-list`
2. The output from `find DIR -type d | wc -l` where DIR is each directory from the `watchman watch-list` output
3. The output from `find DIR -type f | wc -l` where DIR is each directory from the `watchman watch-list` output
4. The output from `find DIR -type l | wc -l` where DIR is each directory from the `watchman watch-list` output
5. The contents of any `.watchmanconfig` files that are present in each of the watched dirs

Please describe the workload in each of the dirs: are there directories where many randomly named files are created and removed continually?  Are there directories containing many build outputs?

Once you've gathered the above, please then try running:

`watchman debug-ageout DIR 0`, where `DIR` is one of your watched dirs.  If you see the memory usage reduce then the cause is that watchman is holding on to a large number of aged file entries.

You can tune how often watchman will look for and remove these nodes:

https://facebook.github.io/watchman/docs/config.html#gc_age_seconds
https://facebook.github.io/watchman/docs/config.html#gc_interval_seconds

I may be able to make other suggestions once you've provided the information above.

**Comment by wez** [comment]
I'm going to close this out due to lack of feedback.  Please re-open this with the requested information when you have a chance to come back to it.

**Comment by ggamel** [comment]
@wez Hi, Wez, I managed to experience this issue again and determined it was from Atom.app w/watchman (https://atom.io) saving open projects. For my situation, watchman had 3 projects (all directories I'd previously opened in Atom) that were actually no longer open in Atom. It was hammering my CPU, though, despite having closed Atom entirely. Not entirely sure how to resolve this issue.

**Comment by wez** [comment]
@ggamel if you could provide the information I asked for above, that would help.  Right now I don't really know anything about the size and shape of those directories or the kind of workload (is it a C/C++ project? Node with lots of npm modules?).

The watchman logs would also be useful to look at https://facebook.github.io/watchman/docs/troubleshooting.html#where-are-the-logs

**Comment by ggamel** [comment]
Sure thing. I was not sure if the Atom involvement information would matter. Will provide that info for you a bit later.

**Comment by ggamel** [comment]
@wez Discovered the root cause of all my pain: https://github.com/facebook/nuclide/issues/84

Turns out Nuclide was causing the watchman process to hammer the CPU. Using Atom 1.9.9 and Nuclide 0.163.0, the watchman issue occurred just as [described here](https://github.com/facebook/nuclide/issues/84#issuecomment-212408027).

For now I've uninstalled Nuclide. If watchman issues persist, I will revisit. Thanks.
